[iOS] JK의 피드백을 반영하였습니다.

ehgud0670 opened the pull request:
###  pull request #119 에 해주신 JK의 피드백을 반영하였습니다. 
> Issue #160 을 구현하였습니다.

* https://github.com/codesquad-member-2020/airbnb-02/pull/119#discussion_r434270295 => 네, lastPathComponent 라는 매개변수 명은 제한적이라 path 로 변경하였습니다.
* https://github.com/codesquad-member-2020/airbnb-02/pull/119#discussion_r434270470 => 네, Equatable은 테스트를 위해 채택한 것이라 채택을 해제(?)하였습니다.
* https://github.com/codesquad-member-2020/airbnb-02/pull/119#discussion_r434271010 => 네,  성공 핸들러와 실패 핸들러로 나누긴 했는데 그 이후에 조치를 취하지는 않았습니다.
* https://github.com/codesquad-member-2020/airbnb-02/pull/119#discussion_r434271563 => 네, 각 메소드가 서로 관련이 없어 보여 프로토콜을 분리하였습니다. 
* https://github.com/codesquad-member-2020/airbnb-02/pull/119#discussion_r434272425 => 아 넵, 중간에 핸들러만 바뀌는 경우를 생각하지 못했습니다... 따라서 한 요청당 한 핸들러를 설정하도록 리팩토링하였습니다. 

* https://github.com/codesquad-member-2020/airbnb-02/pull/119#discussion_r434272555 => RoomImageUseCase 가 노티를 보내는 것이 아니라 해당 이미지의 룸 id를 핸들러로 보내 해당 룸의 셀을 업데이트 하는 구조로 변경하였습니다. 
사실 이미지를 다운로드 받을때 원래의 해당 Room 모델 객체가 바뀌는 것이 있다면 뷰모델에게 알리고 뷰모델이 노티를 보내서 해당 룸의 셀만 업데이트하는 것이 좋다고 판단하는데 이미지를 다운로드 받아도 현재 **Room 모델 객체가 바뀌는 것이 없습니다.** 
<br>이미지가 다운로드 된다 해도 Room 모델은 원래부터 image에 대한 URLString을 가지고 있어서 변할게 없고, 이미지가 저장돼있는 폴더의 destinationURL를 굳이 Room 모델이 프로퍼티로 가지고 있을 필요가 없다는 판단에 그냥 이미지가 다운로드되면 바로 해당 id의 셀을 업데이트 하도록 했습니다.
<br> 이 구조가 맞을까요..? 궁금합니다 🧐
* https://github.com/codesquad-member-2020/airbnb-02/pull/119#discussion_r434272897 => 넵 애초에 생성할때 도메인과 디렉토리를 주입하고 모든 메소드들을 인스턴스 메소드로 변경하였습니다. 이러면 하나의 객체가 하나의 위치(도메인, 디렉토리)에 대해서만 역할을 할 수 있어 객체지향적이라 이전 코드보다 더 좋은 것 같습니다. 감사합니다. 
* https://github.com/codesquad-member-2020/airbnb-02/pull/119#discussion_r434273287 => 네, 말씀해주신대로 큐는 default값으로 nil를 할당하고 block에서 메인큐에다 뷰를 렌더링하는 코드를 집어넣었습니다. 
근데  말씀해주신것처럼 큐를 메인큐로 했을 때 교착상태가 일어날지는 잘 모르겠습니다..  [애플 문서][addObserver]를 보면 큐가 nil인 경우 posting thread가 동기적으로 block를 처리한다는데, 이것처럼 큐를 메인큐로 등록했을때 posting thread가 메인 스레드인 경우에도 block을 메인 스레드가 처리하게 되는 것이므로 정상 작동한다고 할것 같은데 잘 모르겠네용..
* https://github.com/codesquad-member-2020/airbnb-02/pull/119#discussion_r434274149 => 네 뷰모델이 observing하여 특정 값에 의존하는 구조를 제거하였습니다.
* https://github.com/codesquad-member-2020/airbnb-02/pull/119#discussion_r434274539 => 넵, 메소드로 처리하였습니다. 
* https://github.com/codesquad-member-2020/airbnb-02/pull/119#discussion_r434278312 => 네, 해당 객체를 성공쿼리스텁 객체라고 이름 바꾸었습니다. Request 에 대한 실패 메소드 의 경우는 적당한게 떠오르지 않아 구현하지 않았습니다. 
* https://github.com/codesquad-member-2020/airbnb-02/pull/119#discussion_r434278535 => 네 무엇때문에 실패하는지 명시하였습니다. 
* https://github.com/codesquad-member-2020/airbnb-02/pull/119#discussion_r434278779 => 넵 , 디코드 처리만 하면 되지, 따로 비교할 필요는 없다고 생각해 제거하였습니다.

> Slack 질문 
<img width="385" alt="RequestQueue_Comment" src="https://user-images.githubusercontent.com/38216027/84570813-b59a2a80-adca-11ea-9310-6b2a3844f9ca.png">
=> 이 피드백을 반영하기 위해 각 UseCase 에 세마포어 객체를 두어서 동시에 비동기 작업이 10개만 진행되도록 리팩토링했는데 이게 맞는 방법인지 궁금하고, 좀 더 default 한 방법이 있는지 궁금합니다... 😅

[addObserver]: https://developer.apple.com/documentation/foundation/notificationcenter/1411723-addobserver

ehgud0670 committed:
[#160] refactor: 바뀐 모델 데이터에 따라 테스트 관련 코드 변경 및
		  바뀐 모델 이름에 따라 bnb에서 room으로 이름 수정

ehgud0670 committed:
[#160] refactor: 각각의 요청들마다 핸들러로 모델 데이터를 보내도록 수정
	이유: 모든 요청이 동일한 핸들러를 사용하는 구조는 중간에 핸들러가 바뀌는 경우 다른 요청들에 악영향을 줄 수 있는 안좋은 구조이다.
	      따라서 다른 요청에 영향을 주지 않도록 하기 위해 각 요청마다 핸들러를 보내는 구조로 바꾼다.

ehgud0670 committed:
[#160] refactor: Room 에 클로저를 넘겨 image를 처리하는 코드로 수정 및 메소드 분리

ehgud0670 committed:
[#160] refactor: 세마포어를 사용해서 이미지 비동기작업이 동시에 10개만 동작하도록 기능 수정

ehgud0670 committed:
[#160] refactor: 세마포어로 rooms 을 fetch 하는 비동기 작업을 동시적으로 10개로 제한함

ehgud0670 committed:
[#160] refactor: extension으로 나누지 않고 합침
	이유: Equatable 프로토콜을 채택해서 따로 구현하고 있는게 없으므로 합친다.

ehgud0670 committed:
i[#160] refactor: 메소드들을 인스턴스 메소드로 변경 및 매개변수 레이블 이름 수정
	이유: ImageCache 객체의 모든 정적 메소드에 directory, domain 이 동시에 있어서
		이럴거면 인스턴스 메소드로 변경하고 애초에 ImageCache 객체 생성시 directory, domain값을 주는게
		인스턴스 객체가 자신의 역할만 하는 것이고, 응집성이 있다고 생각하여 변경하였다.

	       매개변수 레이블로 lastPathComponent라는 이름은 너무 제한적이라 path 로 변경하였다.

ehgud0670 committed:
[#160] refactor: 메인 스레드에서 옵저버를 등록하는게 옳지 않아 nil로 설정

ehgud0670 committed:
[#160] refactor: 둘의 메소드가 서로 관련이 없어 프로토콜을 Dispatcher, Downloader로 분리

ehgud0670 committed:
[#160] feat: session을 한번 더 wrapping 하는게 의미없다고 판단해 삭제

ehgud0670 committed:
[#160] style: 미처 수정하지 않은 변수명을 수정함

ehgud0670 committed:
[#160] refactor: RoomImageUseCase로 변경하여 roomID를 핸들러에 보내서
		이미지 업데이트 하도록 기능 수정함

ehgud0670 committed:
[#160] refactor: 들여쓰기를 줄이기 위해 파라미터를 Room타입으로 변경 및
		 RoomImageUseCase 가 이미지를 다운로드 하는것이므로 메소드명을 download라고 변경

ehgud0670 committed:
[#160] refactor: execute 메소드에서 completionHandelr(성공), failureHandler(실패)로 분리
		  테스트 코드도 변경에 따라 리팩토링함

ehgud0670 committed:
[#160] refactor: download 메소드에서 completionHandler와 failuerHandler로 분리

ehgud0670 committed:
[#160] refactor: 해당 객체는 Stub이므로 Stub으로 이름 변경 및 download 메소드 제거

ehgud0670 committed:
[#160] refactor: decode만 잘되는지 테스트하는 메소드로 코드 수정 및 Equatable 채택안하도록 함

	이유: 테스트하는 객체의 역할은 디코드를 처리하는 것이라 해당 데이터를 비교하는 것은 필요없는 행동이라 생각했기 떼문이다.
	      또 Equatable을 채택하는 것은 이전 테스트 코드를 위함이지, 프로덕션 코드를 위한 것은 아니라 삭제한다. (테스트만을 위한 코드는 좋지 않다)

ehgud0670 committed:
[#160] style: 실패 테스트가 무엇때문에 실패하는지 더 구체적으로 보이도록 메소드 이름 변경 및
		테스트 메소드를 한글로 수정함

ehgud0670 committed:
[#160] style: 해당 객체가 유효한 쿼리를 가진 Stub객체라 이름 변경하고,
	이에 따라 메소드 명도 변경함

ehgud0670 committed:
[#160] style: 해당 에러는 URL이 아닌 URLString 이 잘못된 것이므로 이름 변경

godrm commented on iOS/Airbnb/Airbnb/Networking/Mocking/RoomsSuccessStub.swift:
```diff
@@ -0,0 +1,20 @@
+        completionHandler: @escaping (Data?, URLResponse?) -> (),
+        failureHandler: @escaping (URLResponse?, Error?) -> ()
+    ) throws {
+        guard let jsonData = Data.readJSON(for: "RoomsData") else { return }
```
데이터 없어도 return 아니라 completionHandler를 호출해줘야 하는거 아닌가요? 실패했는지 인지 할 수가 없습니다.

ehgud0670 replied:
넵 데이터가 없는 경우 failureHandler에 DataError.notFound를 넣도록 하였습니다.

godrm commented on iOS/Airbnb/Airbnb/UseCases/RoomImageUseCase.swift:
```diff
@@ -0,0 +1,52 @@
+    private let imageCache = ImageCache()
+    private let networkDownloader: NetworkDownloader
+    private let urlsQueue = DispatchQueue(label: "urls.serial.queue")
+    private let semaphore = DispatchSemaphore(value: 10)
```
10개라는 숫자는 어떤 기준에서 넣은건가요? 11개는 안되고, 9개도 안되고 10개여야 하는건가요?
다른 더 정확한 방법이 있는 건 아닙니다. 어떤 작업에 대한 동시성이 얼마나 효과적인지는 엔지니어가 실험을 통해서 검증하는 수 밖에 없습니다. 
세모포어를 이용한 방법도 좋은 선택 중에 하나네요. 10개씩 실행되는 것을 확인했나요?

ehgud0670 replied:
1. 음 넵, 말씀해주셔서 10개씩 실행되는지는 오늘 확인하였습니다. 이게 더 효율적인지는 잘 모르겠네요.. 비교를 어떻게 해야되는지 아직은 잘 모르겠습니다.
2. RoomUseCase 는 연달아 request 하는 객체가 아니라서 굳이 세마포어가 필요없다고 판단하였고, 이 객체의 세마포어 객체는 삭제하였습니다.

godrm replied:
세마포어 객체를 제거하면 원하는 개수만큼 조정할 수 없지 않나요?

ehgud0670 replied:
넵, 근데 ImagUseCase 는 이미지를 연달아 다운받을 상황이 충분히 있어서 세마포어 객체를 유지하였고, RoomUseCase는 요청을 많이 할 상황이 없을 것같아 세마포어 객체를 제거하였습니다. 
하지만, 안정성을 고려한다면 RoomUseCase도 요청이 많을 것을 대비해 세마포어로 조절해주는게 맞는건가요?? 🤷‍♂️

ehgud0670 replied:
피드백을 고려하여 git revert을 이용해 세마포어가 있는 이전 상태의 코드로 되돌렸습니다.

godrm replied:
당연히 어떤 UseCase냐에 따라 다를 수 있겠네요. 동시에 여러 요청을 처리할 필요가 없으면 세마포어도 필요없겠습니다.

godrm commented on iOS/Airbnb/Airbnb/ViewModels/RoomViewModels.swift:
```diff
@@ -45,15 +46,19 @@ extension RoomViewModels: UICollectionViewDataSource {
+            guard let image = imageCache.read(path: url.lastPathComponent) else { return }
             
             cell.imagePagingView.insert(at: index, image: image)
             index += 1
```
repeatImages 에서 index도 넘겨주면 좋았을 것 같네요. urlString 도 String 아니라 URL을 넘겨주면 더 간단할 것 같습니다. 
근데 이미지 캐시에 물어볼 때도 lastPathComponent 만 있으면 이미지를 다운로드하고 찾을 수 있나요? 전체가 다 필요한거 아닌가요?
경로 중간만 다르고 이미지 파일명이 같을 수도 있는거 아닌가요?

ehgud0670 replied:
네 좋은 말씀 감사합니다. 👍
중간 경로를 생각하지 못했는데 경로의 중간만 다르고 이미지 파일명이 같은 경우가 충분히 있을 것 같습니다. 
따라서 모든 경로가 보여질수 있게 underscorePath 연산 프로퍼티를 만들어 적용하였습니다.

godrm commented on iOS/Airbnb/AirbnbTests/SessionTests.swift:
```diff
@@ -10,29 +10,27 @@ import XCTest
+                XCTAssertNotNil(data)
+                let httpResponse = response as! HTTPURLResponse
+                XCTAssertEqual(httpResponse.statusCode, 200)
+        }, failureHandler: { response, error in
```
저는 sessionMock을 테스트하는 이 테스트가 이상하다고 생각해요.
mock 객체 자체가 구현이 됐는지 테스트하는 게 아니라, 
우리가 테스트하려는 객체가 있고, 그 객체가 혼자서 동작하지 못하기 때문에 mock 객체를 넘겨줘서 진짜처럼 동작하는 걸 검증해야 합니다. A -> Mock에 요청해서 A가 제대로 동작하는지 검증해야 하지, Mock 자체가 어떻게 동작하는지 보자는 게 아닙니다.

ehgud0670 replied:
네 맞는 말씀입니다. 그래서 생각을 해봤는데 프로토콜 Dispatcher 의 excute 메소드를 테스트를 할 수는 없는거죠?
Session은 Alamofire 객체이므로 해당 메소드는 제가 테스트하는 것이 아니라고 들었습니다. 
따라서 제가 작성한 protocol NetworkDispatcher의 execute메소드를 테스트 하려는데 사실 추상 타입의 메소드를 테스트 할 수 없다고 생각해서요..!

godrm replied:
프로토콜을 채택한 객체의 execute 메소드를 테스트하는 게 아니지 않을까요? 그러면 채택한 모든 객체를 다 테스트할 수 있을까요?
프로토콜을 Mock으로 구현하고 그 메소드를 호출하는 객체를 테스트하는 게 맞는 방향이 아닐까요? 
프로토콜을 채택한 객체는 프로토콜에 선언한 메소드가 아니더라도 그냥 테스트하면 되는데 Mock 객체는 할 필요가 없다는 의미에요.

ehgud0670 replied:
넵 감사합니다! 
* 그래서 Session 객체는 그냥 생성시키고 안에있는 URLProtocol 관련해서만 URLProtocolSuccessStub객체를 할당한뒤에 성공하는 테스트를 하였습니다.
* 그리고 Request을 채택한 RequestInvalidURLStub객체를 이용해 InvalidURL로 인한 실패 테스트도 작성하였습니다.

godrm reviewed (commented):
다음 내용을 확인해보세요.

ehgud0670 committed:
[#160] refactor: 1. repeatImages에 URL, index를 핸들러로 내보내도록 리팩토링
		 2. data를 파일로 부터 못받을 경우 failureHandler로 에러를 보내도록 기능 수정

		1 의 이유: 상위객체는 단순하게, 하위 객체는 복잡하게!

ehgud0670 committed:
[#160] refactor: RoomUseCase 의 세마포어 객체 삭제

	이유: roomUseCase는 연달아 request를 하는 객체가 아니므로
		세마포어 자체가 필요없다고 생각한다. 따라서 삭제한다.

ehgud0670 committed:
[#160] fix: url의 중간 경로를 나타내기 위해 연산 프로퍼티 구현 및 적용

		이유: 이미지 중에 이미지 이름만 같고 url의 중간 경로는 다른 경우는 확실히 있다.
		      따라서 중간 경로를 고려하지 않으면 이미지 이름이 중복되어 잘못 찾게 되는 버그가 생길 수 있으므로
		     중간 경로도 모두 내포할 수 있는 연산 프로퍼티를 만들어 해결하였다.

ehgud0670 committed:
[#160] refactor: session에 mock라는 이름 제거, 안의 URLProtocol 프로퍼티만
		  Mock로 설정하고, Session의 execute 메소드를 테스트함

ehgud0670 committed:
[#160] refactor: 언래핑 실패가 테스트 실패로 표시되도록 ! 삭제 및 throws 붙임

ehgud0670 committed:
[#160] feat: invalidURL로 인한 실패하는 테스트 추가

ehgud0670 committed:
Revert "[#160] refactor: RoomUseCase 의 세마포어 객체 삭제"

This reverts commit f7bdd6313e850b2a64bbb78201db2d7e236081b2.

ehgud0670 commented:
다음 스텝을 위해 머지하겠습니다. 👍

godrm commented:
수고하셨습니다.
